hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java

```diff
@@ -48,20 +51,45 @@
  */
 public abstract class SpecificPartitionsImapReaderPms<F extends CompletableFuture, B, R>
         extends LocalProcessorMetaSupplier<F, B, R> {
+    private final List<List<Expression<?>>> requiredPartitionsExprs;
     private transient int[] partitionsToScan;
     private transient Map<Address, int[]> partitionAssignment;
 
     private SpecificPartitionsImapReaderPms(
-            final BiFunctionEx<HazelcastInstance, InternalSerializationService, Reader<F, B, R>> readerSupplier) {
+            final BiFunctionEx<HazelcastInstance, InternalSerializationService, Reader<F, B, R>> readerSupplier,
+            @Nullable final List<List<Expression<?>>> requiredPartitionsExprs) {
         super(readerSupplier);
+        this.requiredPartitionsExprs = requiredPartitionsExprs;
     }
 
     @Override
     public void init(@Nonnull Context context) throws Exception {
         super.init(context);
-        Set<Integer> partitions = context.jobConfig().getArgument(KEY_REQUIRED_PARTITIONS);
-        if (partitions != null) {
-            partitionsToScan = partitions.stream().sorted().mapToInt(i -> i).toArray();
+        if (requiredPartitionsExprs != null) {
+            List<Integer> partitionsToScanList = new ArrayList<>();
+            partitionsToScan = new int[requiredPartitionsExprs.size()];
+            int j = 0;
+
+            HazelcastInstance hazelcastInstance = context.hazelcastInstance();
+            ExpressionEvalContext eec = ExpressionEvalContext.from(context);
+            for (List<Expression<?>> requiredPartitionsExpr : requiredPartitionsExprs) {
+                Object[] partitionKeyComponents = new Object[requiredPartitionsExpr.size()];
+                int i = 0;
+                for (Expression<?> expression : requiredPartitionsExpr) {
+                    partitionKeyComponents[i++] = expression.eval(null, eec);
+                }
+
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
+                        StrategyUtil.constructKey(partitionKeyComponents)
+                );
+                if (partition == null) {
+                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
+                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
+                    break;
```

Comment:
`partitionsToScan` will be incomplete in this case instead of containing all patitions.

Reply:
this is not fixed